runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -163,6 +165,84 @@ public void optimisticRollbackUnsafe() {
         modifyingContext = null;
     }
 
+    /** Roll the object back to the supplied version if possible.
+     * This function may roll back to a point prior to the requested version.
+     * Otherwise, throws a NoRollbackException.
+     *
+     * Unsafe, requires that the caller has acquired a write lock.
+     *
+     * @param  version              The version to rollback to.
+     * @throws NoRollbackException  If the object cannot be rolled back to
+     *                              the supplied version.
+     */
+    public void rollbackUnsafe(long version) {
+        // If we're already at or before the given version, there's
+        // nothing to do
+        if (this.version <= version) {
+            return;
+        }
+
+        // If we don't have an undo log, we can't roll back.
+        if (undoLog.size() == 0) {
+            throw new NoRollbackException();
+        }
+
+        while (undoLog.size() > 0) {
+            SMREntry undoRecord = undoLog.pollFirst();
+
+            // Make sure the record is undoable.
+            // This should never happen, but if
+            // for some reason the undo log contains an
+            // undoable entry, clear the log and throw an
+            // exception.
+            if (!undoRecord.isUndoable()) {
+                undoLog.clear();
+                throw new NoRollbackException();
+            }
+
+            // Apply the undo record.
+            undoFunctionMap.get(undoRecord.getSMRMethod())
+                    .doUndo(object, undoRecord.getUndoRecord(),
+                                    undoRecord.getSMRArguments());
+
+            // update the version
+            this.version = undoRecord.getEntry().getGlobalAddress();
+
+            // check if we rolled back to the requested version
+            if (this.version <= version) {
+                return;
+            }
+        }
+
+        throw new NoRollbackException();
+    }
+
+    /** Calculate the number of undo records we need to keep,
+     * possibly cleaning up the undo log, and return whether
+     * we need to keep this undo record.
+     *
+     * @return  True, if an undo record is needed in the undo log,
+     *          False otherwise.
+     */
+    public boolean needUndoRecordUnsafe() {
+
+        // Now get the oldest transaction in the context set.
+        long oldestVersion = TransactionalContext.getOldestSnapshot();
+        // If there are no active transactions, or all active transactions
+        // are after this object's version, we can just drop everything.
+        if (oldestVersion == -1L || oldestVersion > version) {
+            undoLog.clear();
+            return false;
+        }
+
+        // make sure we have at most enough undo entries
+        // to go between this version and the oldest version
+        while (undoLog.size() > version - oldestVersion + 1) {
+            undoLog.pollLast();
```

Comment:
Let me clarify: I am asking whether this while loop can inspect undoRecord.getEntry().getGlobalAddress() in order to decide how far back to keep entries?